File: 040-X/christer/regexp2.py
import re
import time

def ass(a,b):
	if a != b:
		print("assert failure:")
		print(a)
		print("!=")
		print(b)

def fetchText(filename):
	with open(filename) as f:
		return f.read()

start = time.time_ns()

################# general

any = '.+?' # zero or one character, lazy

brackets = lambda s : f'<{s}>'
ass(r'<x>', brackets('x'))

opt = lambda s : f'{s}?'
ass(r'x?', opt('x'))

group = lambda s : f'({s})'
ass(r'(x)', group('x'))

q = lambda s: ''.join(["\\" + ch for ch in s])
ass(r'\(\*', q('(*'))

nl = q('n')
ass(r'\n', q('n'))

################## application specific

pascalComment = lambda s : q('(*') + s + q('*)')
ass(r'\(\*x\*\)', pascalComment('x'))

group1 = group(' ' + brackets(any) + opt('*') + ':')
ass(r'( <.+?>*?:)', group1)

group2 = group(any)
ass(r'(.+?)', group2)

rex = nl + pascalComment(group1 + opt(nl) + group2 + nl)
ass(r'\n\(\*( <.+?>*?:)\n?(.+?)\n\*\)', rex)

regex = re.compile(rex, re.DOTALL)
text = fetchText('../xal.pas')

matches = re.findall(regex,text)
for (name,body) in matches:
	print(name)
	print(body)

print(len(matches), 'matches')
print((time.time_ns()-start)/10**6,'ms')

'''
Borde gå att få se detta i utvecklingsmiljön:
Man redigerar trädet på vänster sida
Högra sidan skapas automatiskt
	(Är den öht intressant utom för debug?)

(rex                  # \n\(\*( <.+?>*?:)\n?(.+?)\n\*\)
	nl                # \n
	(pascalComment    #   \(\*( <.+?>*?:)\n?(.+?)\n\*\)
		(q            #   \(\*
			'(*'      #   '(*'  
    	(group        #       ( <.+?>*?:)
			' '       #        .
			(brackets #         <.+?>
				any   #          .+?
			(opt      #              *?
				'*'   #              *
			':'       #                :
		(opt          #                  \n?
			nl        #                  \n            
		(group        #                     (.+?)
			any       #                      .+?
		nl            #                          \n
		(q            #                            \*\)
			'*)'      #                            '*)'
'''

File: 040-X/christer/regexp2.clj
(ns user (:require [clojure.string :as str]))

(defn ass [a,b] (if (= a b) "ok" (str "Assert failure: " a " != " b)))

;;;;;;;;;;;;;;;;;; general

(def any ".+?") ;; zero or one character, lazy
(ass ".+?" any)

(defn brackets [& s] (str "<" (str/join s) ">"))
(ass "<x>" (brackets "x"))

(defn opt [s] (str s "?"))
(ass "x?" (opt "x"))

(defn group [& s] (str "(" (str/join s) ")"))
(ass "(x)" (group "x"))

(defn q [s] (str/join "" (map #(str "\\" %) s)))
(ass "\\(\\*" (q "(*"))

(def nl (q "n"))
(ass "\\n" nl)

;;;;;;;;;;;;;;;;;;;;;;; application specific

(defn pascalComment [& s] (str nl (q "(*") (str/join s) (q "*)")))
(ass "\\n\\(\\*x\\*\\)"  (pascalComment "x"))

;; Man redigerar trädet på vänster sida
;; Högra sidan skapas automatiskt (Är den öht intressant utom för debug?)

(pascalComment      ;; \\n\\(\\*\\n(<.+?*?>):\\n?(.+?)\\n\\*\\)
 (group             ;;             (<.+?*?>)
  (brackets         ;;              <.+?*?>
   any              ;;               .+?
   (opt "*"))       ;;                  *?
  ":"               ;;                      :
  (opt nl)          ;;                       \\n?
  (group            ;;                           (.+?)
   any)             ;;                            .+?
  nl))              ;;                                \\n


